In [1]:
import pandas as pd

df = pd.read_csv("https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv")

In [2]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [3]:
df.tail(3)

PassengerId  Survived  Pclass                                      Name  \
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch      Ticket   Fare Cabin Embarked  
888  female   NaN      1      2  W./C. 6607  23.45   NaN        S  
889    male  26.0      0      0      111369  30.00  C148        C  
890    male  32.0      0      0      370376   7.75   NaN        Q

In [4]:
df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [5]:
df.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

The most important thing is to have knowledge on rows and columns and what they mean. eg the above dataset is titanic dataset

* passenger id  - booking id
* Survived - 1(yes)  ,  0(no)
* Pclass - 1st or 2nd or 3rd class

In [6]:
df.describe()  # statistical analysis on dataset

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

Note: The above function is only giving info on numerical data and not categorical data as statistics only works with numbers. name and sex is categorical. Hence, it is not present in the o/p of the output of the above cell.

In [7]:
# Consider you want to perform statistics on categorical data

df.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

In [8]:
df[["Name", "Sex", "Ticket", "Cabin", "Embarked"]]

Name     Sex  \
0                              Braund, Mr. Owen Harris    male   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female   
2                               Heikkinen, Miss. Laina  female   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female   
4                             Allen, Mr. William Henry    male   
..                                                 ...     ...   
886                              Montvila, Rev. Juozas    male   
887                       Graham, Miss. Margaret Edith  female   
888           Johnston, Miss. Catherine Helen "Carrie"  female   
889                              Behr, Mr. Karl Howell    male   
890                                Dooley, Mr. Patrick    male   

               Ticket Cabin Embarked  
0           A/5 21171   NaN        S  
1            PC 17599   C85        C  
2    STON/O2. 3101282   NaN        S  
3              113803  C123        S  
4              373450   NaN        S  
..                ...   ...      ...  
886            211536   NaN        S  
887            112053   B42        S  
888        W./C. 6607   NaN        S  
889            111369  C148        C  
890            370376   NaN        Q  

[891 rows x 5 columns]

But this is possible for less number of columns. If you have large number of columns then extracting categorical data like this is not efficient

In [9]:
# Solution

type(df.dtypes)

pandas.core.series.Series

This means,  passengerid, survived, etc are all index

In [10]:
df.dtypes == "object"

PassengerId    False
Survived       False
Pclass         False
Name            True
Sex             True
Age            False
SibSp          False
Parch          False
Ticket          True
Fare           False
Cabin           True
Embarked        True
dtype: bool

Here, passengerid, survived, etc are all index

In [11]:
df.dtypes[df.dtypes == "object"]

Name        object
Sex         object
Ticket      object
Cabin       object
Embarked    object
dtype: object

Hence, in this way, only categorical data can be extracted.

In [12]:
df.dtypes[df.dtypes == "object"].index

Index(['Name', 'Sex', 'Ticket', 'Cabin', 'Embarked'], dtype='object')

In [13]:
# Extracting categorical columns from dataset

df[ df.dtypes[df.dtypes == "object"].index ]

Name     Sex  \
0                              Braund, Mr. Owen Harris    male   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female   
2                               Heikkinen, Miss. Laina  female   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female   
4                             Allen, Mr. William Henry    male   
..                                                 ...     ...   
886                              Montvila, Rev. Juozas    male   
887                       Graham, Miss. Margaret Edith  female   
888           Johnston, Miss. Catherine Helen "Carrie"  female   
889                              Behr, Mr. Karl Howell    male   
890                                Dooley, Mr. Patrick    male   

               Ticket Cabin Embarked  
0           A/5 21171   NaN        S  
1            PC 17599   C85        C  
2    STON/O2. 3101282   NaN        S  
3              113803  C123        S  
4              373450   NaN        S  
..                ...   ...      ...  
886            211536   NaN        S  
887            112053   B42        S  
888        W./C. 6607   NaN        S  
889            111369  C148        C  
890            370376   NaN        Q  

[891 rows x 5 columns]

In [14]:
df[ df.dtypes[df.dtypes == "object"].index ].describe()

Name   Sex  Ticket    Cabin Embarked
count                       891   891     891      204      889
unique                      891     2     681      147        3
top     Braund, Mr. Owen Harris  male  347082  B96 B98        S
freq                          1   577       7        4      644

This is analysis of categorical data

Here, top is somewhat similar to mode and freq shows number of times top has appeared

In [15]:
# Filter out columns where data type is integer

df.dtypes[df.dtypes == "int64"]

PassengerId    int64
Survived       int64
Pclass         int64
SibSp          int64
Parch          int64
dtype: object

In [16]:
df.dtypes[df.dtypes == "int64"].index

Index(['PassengerId', 'Survived', 'Pclass', 'SibSp', 'Parch'], dtype='object')

In [17]:
df[df.dtypes[df.dtypes == "int64"].index]

PassengerId  Survived  Pclass  SibSp  Parch
0              1         0       3      1      0
1              2         1       1      1      0
2              3         1       3      0      0
3              4         1       1      1      0
4              5         0       3      0      0
..           ...       ...     ...    ...    ...
886          887         0       2      0      0
887          888         1       1      0      0
888          889         0       3      1      2
889          890         1       1      0      0
890          891         0       3      0      0

[891 rows x 5 columns]

# Other option

In [18]:
import numpy as np

df.select_dtypes(include=np.int64)

PassengerId  Survived  Pclass  SibSp  Parch
0              1         0       3      1      0
1              2         1       1      1      0
2              3         1       3      0      0
3              4         1       1      1      0
4              5         0       3      0      0
..           ...       ...     ...    ...    ...
886          887         0       2      0      0
887          888         1       1      0      0
888          889         0       3      1      2
889          890         1       1      0      0
890          891         0       3      0      0

[891 rows x 5 columns]

In [19]:
df.select_dtypes(include=np.number).columns

Index(['PassengerId', 'Survived', 'Pclass', 'Age', 'SibSp', 'Parch', 'Fare'], dtype='object')

In [20]:
df.select_dtypes(include=np.number).columns.tolist()

['PassengerId', 'Survived', 'Pclass', 'Age', 'SibSp', 'Parch', 'Fare']

In [21]:
df.select_dtypes(include=np.float64)

Age     Fare
0    22.0   7.2500
1    38.0  71.2833
2    26.0   7.9250
3    35.0  53.1000
4    35.0   8.0500
..    ...      ...
886  27.0  13.0000
887  19.0  30.0000
888   NaN  23.4500
889  26.0  30.0000
890  32.0   7.7500

[891 rows x 2 columns]

In [22]:
# Using our method. ( without using numpy)

df[df.dtypes[df.dtypes == "float64"].index]

Age     Fare
0    22.0   7.2500
1    38.0  71.2833
2    26.0   7.9250
3    35.0  53.1000
4    35.0   8.0500
..    ...      ...
886  27.0  13.0000
887  19.0  30.0000
888   NaN  23.4500
889  26.0  30.0000
890  32.0   7.7500

[891 rows x 2 columns]

In [23]:
df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [24]:
df[['Survived', 'Pclass']]

Survived  Pclass
0           0       3
1           1       1
2           1       3
3           1       1
4           0       3
..        ...     ...
886         0       2
887         1       1
888         0       3
889         1       1
890         0       3

[891 rows x 2 columns]

In [25]:
# Extract same columns but rows from 5 to 12

df[['Survived', 'Pclass']][5:13]

Survived  Pclass
5          0       3
6          0       1
7          0       3
8          1       3
9          1       2
10         1       3
11         1       1
12         0       3

It is nothing but slicing operation

In [26]:
# Adding a column to a dataframe

print(df.head(2))

# Add a new column name

df['new_col'] = 0
df.head(2)          # new column will be added at last

   PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   

   Parch     Ticket     Fare Cabin Embarked  
0      0  A/5 21171   7.2500   NaN        S  
1      0   PC 17599  71.2833   C85        C  


PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   

   Parch     Ticket     Fare Cabin Embarked  new_col  
0      0  A/5 21171   7.2500   NaN        S        0  
1      0   PC 17599  71.2833   C85        C        0

In [27]:
# Adding column which contains sum of passengerid and Pclass

df['new_col1'] =  df['PassengerId'] + df['Pclass']
df.head(2)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   

   Parch     Ticket     Fare Cabin Embarked  new_col  new_col1  
0      0  A/5 21171   7.2500   NaN        S        0         4  
1      0   PC 17599  71.2833   C85        C        0         3

In [28]:
df['Pclass']

0      3
1      1
2      3
3      1
4      3
      ..
886    2
887    1
888    3
889    1
890    3
Name: Pclass, Length: 891, dtype: int64

In [30]:
# Suppose we want to find out the categories in a cloumn say Pclass

pd.Categorical(df['Pclass'])

[3, 1, 3, 1, 3, ..., 2, 1, 3, 1, 3]
Length: 891
Categories (3, int64): [1, 2, 3]

Hence, in Pclass, we have 3 categories which are all integers -> 1,2,3

In [35]:
pd.Categorical(df['Survived'])

[0, 1, 1, 1, 0, ..., 0, 1, 0, 1, 0]
Length: 891
Categories (2, int64): [0, 1]

Hence, in Survived, we have 2 categories which are all integers -> 0,1

In [37]:
pd.Categorical(df['Cabin'])

[NaN, 'C85', NaN, 'C123', NaN, ..., NaN, 'B42', NaN, 'C148', NaN]
Length: 891
Categories (147, object): ['A10', 'A14', 'A16', 'A19', ..., 'F38', 'F4', 'G6', 'T']

Hence, in Cabin column, there are 147 Categories which are all objects

In [38]:
# Now, to find unique values in a column .i.e categories only, we can use df[].unique()

df['Pclass'].unique()

array([3, 1, 2])

In [40]:
df['Survived'].unique()

array([0, 1])

In [41]:
df['Cabin'].unique()

array([nan, 'C85', 'C123', 'E46', 'G6', 'C103', 'D56', 'A6',
       'C23 C25 C27', 'B78', 'D33', 'B30', 'C52', 'B28', 'C83', 'F33',
       'F G73', 'E31', 'A5', 'D10 D12', 'D26', 'C110', 'B58 B60', 'E101',
       'F E69', 'D47', 'B86', 'F2', 'C2', 'E33', 'B19', 'A7', 'C49', 'F4',
       'A32', 'B4', 'B80', 'A31', 'D36', 'D15', 'C93', 'C78', 'D35',
       'C87', 'B77', 'E67', 'B94', 'C125', 'C99', 'C118', 'D7', 'A19',
       'B49', 'D', 'C22 C26', 'C106', 'C65', 'E36', 'C54',
       'B57 B59 B63 B66', 'C7', 'E34', 'C32', 'B18', 'C124', 'C91', 'E40',
       'T', 'C128', 'D37', 'B35', 'E50', 'C82', 'B96 B98', 'E10', 'E44',
       'A34', 'C104', 'C111', 'C92', 'E38', 'D21', 'E12', 'E63', 'A14',
       'B37', 'C30', 'D20', 'B79', 'E25', 'D46', 'B73', 'C95', 'B38',
       'B39', 'B22', 'C86', 'C70', 'A16', 'C101', 'C68', 'A10', 'E68',
       'B41', 'A20', 'D19', 'D50', 'D9', 'A23', 'B50', 'A26', 'D48',
       'E58', 'C126', 'B71', 'B51 B53 B55', 'D49', 'B5', 'B20', 'F G63',
       'C62 C64',

In [44]:
len( df['Cabin'].unique() )

148

In [50]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  new_col  new_col1  
0      0         A/5 21171   7.2500   NaN        S        0         4  
1      0          PC 17599  71.2833   C85        C        0         3  
2      0  STON/O2. 3101282   7.9250   NaN        S        0         6  
3      0            113803  53.1000  C123        S        0         5  
4      0            373450   8.0500   NaN        S        0         8

In [51]:
# Filter out data whose age is greater than 18

df["Age"] > 18

0       True
1       True
2       True
3       True
4       True
       ...  
886     True
887     True
888    False
889     True
890     True
Name: Age, Length: 891, dtype: bool

In [52]:
df[df["Age"] > 18]

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
885          886         0       3   
886          887         0       2   
887          888         1       1   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
885               Rice, Mrs. William (Margaret Norton)  female  39.0      0   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  new_col  new_col1  
0        0         A/5 21171   7.2500   NaN        S        0         4  
1        0          PC 17599  71.2833   C85        C        0         3  
2        0  STON/O2. 3101282   7.9250   NaN        S        0         6  
3        0            113803  53.1000  C123        S        0         5  
4        0            373450   8.0500   NaN        S        0         8  
..     ...               ...      ...   ...      ...      ...       ...  
885      5            382652  29.1250   NaN        Q        0       889  
886      0            211536  13.0000   NaN        S        0       889  
887      0            112053  30.0000   B42        S        0       889  
889      0            111369  30.0000  C148        C        0       891  
890      0            370376   7.7500   NaN        Q        0       894  

[575 rows x 14 columns]

In [57]:
# Filter out data whose age is less than 18

df[  df["Age"] < 18]

PassengerId  Survived  Pclass                                     Name  \
7              8         0       3           Palsson, Master. Gosta Leonard   
9             10         1       2      Nasser, Mrs. Nicholas (Adele Achem)   
10            11         1       3          Sandstrom, Miss. Marguerite Rut   
14            15         0       3     Vestrom, Miss. Hulda Amanda Adolfina   
16            17         0       3                     Rice, Master. Eugene   
..           ...       ...     ...                                      ...   
850          851         0       3  Andersson, Master. Sigvard Harald Elias   
852          853         0       3                  Boulos, Miss. Nourelain   
853          854         1       1                Lines, Miss. Mary Conover   
869          870         1       3          Johnson, Master. Harold Theodor   
875          876         1       3         Najib, Miss. Adele Kiamie "Jane"   

        Sex   Age  SibSp  Parch    Ticket     Fare Cabin Embarked  new_col  \
7      male   2.0      3      1    349909  21.0750   NaN        S        0   
9    female  14.0      1      0    237736  30.0708   NaN        C        0   
10   female   4.0      1      1   PP 9549  16.7000    G6        S        0   
14   female  14.0      0      0    350406   7.8542   NaN        S        0   
16     male   2.0      4      1    382652  29.1250   NaN        Q        0   
..      ...   ...    ...    ...       ...      ...   ...      ...      ...   
850    male   4.0      4      2    347082  31.2750   NaN        S        0   
852  female   9.0      1      1      2678  15.2458   NaN        C        0   
853  female  16.0      0      1  PC 17592  39.4000   D28        S        0   
869    male   4.0      1      1    347742  11.1333   NaN        S        0   
875  female  15.0      0      0      2667   7.2250   NaN        C        0   

     new_col1  
7          11  
9          12  
10         14  
14         18  
16         20  
..        ...  
850       854  
852       856  
853       855  
869       873  
875       879  

[113 rows x 14 columns]

In [61]:
# Find out number of people whose fare was lesser than avg fare

df["Fare"].mean()

32.204207968574636

In [62]:
df[ df["Fare"] < df["Fare"].mean() ]

PassengerId  Survived  Pclass                                      Name  \
0              1         0       3                   Braund, Mr. Owen Harris   
2              3         1       3                    Heikkinen, Miss. Laina   
4              5         0       3                  Allen, Mr. William Henry   
5              6         0       3                          Moran, Mr. James   
7              8         0       3            Palsson, Master. Gosta Leonard   
..           ...       ...     ...                                       ...   
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked  \
0      male  22.0      1      0         A/5 21171   7.2500   NaN        S   
2    female  26.0      0      0  STON/O2. 3101282   7.9250   NaN        S   
4      male  35.0      0      0            373450   8.0500   NaN        S   
5      male   NaN      0      0            330877   8.4583   NaN        Q   
7      male   2.0      3      1            349909  21.0750   NaN        S   
..      ...   ...    ...    ...               ...      ...   ...      ...   
886    male  27.0      0      0            211536  13.0000   NaN        S   
887  female  19.0      0      0            112053  30.0000   B42        S   
888  female   NaN      1      2        W./C. 6607  23.4500   NaN        S   
889    male  26.0      0      0            111369  30.0000  C148        C   
890    male  32.0      0      0            370376   7.7500   NaN        Q   

     new_col  new_col1  
0          0         4  
2          0         6  
4          0         8  
5          0         9  
7          0        11  
..       ...       ...  
886        0       889  
887        0       889  
888        0       892  
889        0       891  
890        0       894  

[680 rows x 14 columns]

In [63]:
len( df[ df["Fare"] < df["Fare"].mean() ]  )

680

Hence, 680 people had a fare of less than the average fare

In [64]:
len( df[ df["Fare"] > df["Fare"].mean() ]  )

211

Hence, 211 people had a fare of more than the average fare

In [65]:
len( df[ df["Fare"] == 0 ] )

15

Hence, 15 people paid 0 Fare

In [70]:
# IF you want to find the names of people who paid 0 Fare

df[ df["Fare"] == 0 ]['Name']

179                 Leonard, Mr. Lionel
263               Harrison, Mr. William
271        Tornquist, Mr. William Henry
277         Parkes, Mr. Francis "Frank"
302     Johnson, Mr. William Cahoone Jr
413      Cunningham, Mr. Alfred Fleming
466               Campbell, Mr. William
481    Frost, Mr. Anthony Wood "Archie"
597                 Johnson, Mr. Alfred
633       Parr, Mr. William Henry Marsh
674          Watson, Mr. Ennis Hastings
732                Knight, Mr. Robert J
806              Andrews, Mr. Thomas Jr
815                    Fry, Mr. Richard
822     Reuchlin, Jonkheer. John George
Name: Name, dtype: object

In [73]:
#  Find number of males on ship

df["Sex"] == "male"

0       True
1      False
2      False
3      False
4       True
       ...  
886     True
887    False
888    False
889     True
890     True
Name: Sex, Length: 891, dtype: bool

In [74]:
df[ df["Sex"] == "male" ]

PassengerId  Survived  Pclass                            Name   Sex  \
0              1         0       3         Braund, Mr. Owen Harris  male   
4              5         0       3        Allen, Mr. William Henry  male   
5              6         0       3                Moran, Mr. James  male   
6              7         0       1         McCarthy, Mr. Timothy J  male   
7              8         0       3  Palsson, Master. Gosta Leonard  male   
..           ...       ...     ...                             ...   ...   
883          884         0       2   Banfield, Mr. Frederick James  male   
884          885         0       3          Sutehall, Mr. Henry Jr  male   
886          887         0       2           Montvila, Rev. Juozas  male   
889          890         1       1           Behr, Mr. Karl Howell  male   
890          891         0       3             Dooley, Mr. Patrick  male   

      Age  SibSp  Parch            Ticket     Fare Cabin Embarked  new_col  \
0    22.0      1      0         A/5 21171   7.2500   NaN        S        0   
4    35.0      0      0            373450   8.0500   NaN        S        0   
5     NaN      0      0            330877   8.4583   NaN        Q        0   
6    54.0      0      0             17463  51.8625   E46        S        0   
7     2.0      3      1            349909  21.0750   NaN        S        0   
..    ...    ...    ...               ...      ...   ...      ...      ...   
883  28.0      0      0  C.A./SOTON 34068  10.5000   NaN        S        0   
884  25.0      0      0   SOTON/OQ 392076   7.0500   NaN        S        0   
886  27.0      0      0            211536  13.0000   NaN        S        0   
889  26.0      0      0            111369  30.0000  C148        C        0   
890  32.0      0      0            370376   7.7500   NaN        Q        0   

     new_col1  
0           4  
4           8  
5           9  
6           8  
7          11  
..        ...  
883       886  
884       888  
886       889  
889       891  
890       894  

[577 rows x 14 columns]

In [78]:
len( df[df["Sex"] == "male" ])

577

Hence, there were 577 males

In [79]:
len( df[df["Sex"] == "female" ])

314

Hence, there were 314 females

In [80]:
# Find passengers with class 1 ticket

df["Pclass"] == 1

0      False
1       True
2      False
3       True
4      False
       ...  
886    False
887     True
888    False
889     True
890    False
Name: Pclass, Length: 891, dtype: bool

In [82]:
df[ df["Pclass"] == 1 ]['Name']


1      Cumings, Mrs. John Bradley (Florence Briggs Th...
3           Futrelle, Mrs. Jacques Heath (Lily May Peel)
6                                McCarthy, Mr. Timothy J
11                              Bonnell, Miss. Elizabeth
23                          Sloper, Mr. William Thompson
                             ...                        
871     Beckwith, Mrs. Richard Leonard (Sallie Monypeny)
872                             Carlsson, Mr. Frans Olof
879        Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)
887                         Graham, Miss. Margaret Edith
889                                Behr, Mr. Karl Howell
Name: Name, Length: 216, dtype: object

In [83]:
len( df[ df["Pclass"] == 1 ] )

216

216 passengers had Class 1 ticket

In [85]:
# Find out how many people survived

len( df[ df["Survived"] == 1 ] )

342

In [86]:
# Find out how many people died

len( df[ df["Survived"] == 0 ] )

549

In [91]:
# Find females who paid fare of more than the avg fare

df["Sex"] == "female"

0      False
1       True
2       True
3       True
4      False
       ...  
886    False
887     True
888     True
889    False
890    False
Name: Sex, Length: 891, dtype: bool

In [92]:
df["Fare"] > df["Fare"].mean()

0      False
1       True
2      False
3       True
4      False
       ...  
886    False
887    False
888    False
889    False
890    False
Name: Fare, Length: 891, dtype: bool

In [95]:
(df["Sex"] == "female") & (df["Fare"] > df["Fare"].mean())

0      False
1       True
2      False
3       True
4      False
       ...  
886    False
887    False
888    False
889    False
890    False
Length: 891, dtype: bool

In [96]:
df[ (df["Sex"] == "female") & (df["Fare"] > df["Fare"].mean()) ]

PassengerId  Survived  Pclass  \
1              2         1       1   
3              4         1       1   
31            32         1       1   
43            44         1       2   
52            53         1       1   
..           ...       ...     ...   
853          854         1       1   
856          857         1       1   
863          864         0       3   
871          872         1       1   
879          880         1       1   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
31      Spencer, Mrs. William Augustus (Marie Eugenie)  female   NaN      1   
43            Laroche, Miss. Simonne Marie Anne Andree  female   3.0      1   
52            Harper, Mrs. Henry Sleeper (Myna Haxtun)  female  49.0      1   
..                                                 ...     ...   ...    ...   
853                          Lines, Miss. Mary Conover  female  16.0      0   
856         Wick, Mrs. George Dennick (Mary Hitchcock)  female  45.0      1   
863                  Sage, Miss. Dorothy Edith "Dolly"  female   NaN      8   
871   Beckwith, Mrs. Richard Leonard (Sallie Monypeny)  female  47.0      1   
879      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0      0   

     Parch         Ticket      Fare Cabin Embarked  new_col  new_col1  
1        0       PC 17599   71.2833   C85        C        0         3  
3        0         113803   53.1000  C123        S        0         5  
31       0       PC 17569  146.5208   B78        C        0        33  
43       2  SC/Paris 2123   41.5792   NaN        C        0        46  
52       0       PC 17572   76.7292   D33        C        0        54  
..     ...            ...       ...   ...      ...      ...       ...  
853      1       PC 17592   39.4000   D28        S        0       855  
856      1          36928  164.8667   NaN        S        0       858  
863      2       CA. 2343   69.5500   NaN        S        0       867  
871      1          11751   52.5542   D35        S        0       873  
879      1          11767   83.1583   C50        C        0       881  

[104 rows x 14 columns]

In [97]:
df[ (df["Sex"] == "female") & (df["Fare"] > df["Fare"].mean()) ]['Name']

1      Cumings, Mrs. John Bradley (Florence Briggs Th...
3           Futrelle, Mrs. Jacques Heath (Lily May Peel)
31        Spencer, Mrs. William Augustus (Marie Eugenie)
43              Laroche, Miss. Simonne Marie Anne Andree
52              Harper, Mrs. Henry Sleeper (Myna Haxtun)
                             ...                        
853                            Lines, Miss. Mary Conover
856           Wick, Mrs. George Dennick (Mary Hitchcock)
863                    Sage, Miss. Dorothy Edith "Dolly"
871     Beckwith, Mrs. Richard Leonard (Sallie Monypeny)
879        Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)
Name: Name, Length: 104, dtype: object

These are names of females who paid fare more than avg fare

In [98]:
len( df[ (df["Sex"] == "female") & (df["Fare"] > df["Fare"].mean()) ])

104

104 females paid fare of more than avg fare

In [99]:
len( df[ (df["Sex"] == "male") & (df["Fare"] > df["Fare"].mean()) ])

107

107 males paid fare of more than avg fare

In [106]:
# Find name of person who paid maximum fare

df["Fare"] == df["Fare"].max()

0      False
1      False
2      False
3      False
4      False
       ...  
886    False
887    False
888    False
889    False
890    False
Name: Fare, Length: 891, dtype: bool

In [107]:
df[ df["Fare"] == df["Fare"].max() ]

PassengerId  Survived  Pclass                                Name  \
258          259         1       1                    Ward, Miss. Anna   
679          680         1       1  Cardeza, Mr. Thomas Drake Martinez   
737          738         1       1              Lesurer, Mr. Gustave J   

        Sex   Age  SibSp  Parch    Ticket      Fare        Cabin Embarked  \
258  female  35.0      0      0  PC 17755  512.3292          NaN        C   
679    male  36.0      0      1  PC 17755  512.3292  B51 B53 B55        C   
737    male  35.0      0      0  PC 17755  512.3292         B101        C   

     new_col  new_col1  
258        0       260  
679        0       681  
737        0       739

In [108]:
df[ df["Fare"] == df["Fare"].max() ]['Name']

258                      Ward, Miss. Anna
679    Cardeza, Mr. Thomas Drake Martinez
737                Lesurer, Mr. Gustave J
Name: Name, dtype: object

These 3 people paid maximum fare

In [113]:
# Find records from row no 0 to 100 with even records only

df.index % 2== 0

array([ True, False,  True, False,  True, False,  True, False,  True,
       False,  True, False,  True, False,  True, False,  True, False,
        True, False,  True, False,  True, False,  True, False,  True,
       False,  True, False,  True, False,  True, False,  True, False,
        True, False,  True, False,  True, False,  True, False,  True,
       False,  True, False,  True, False,  True, False,  True, False,
        True, False,  True, False,  True, False,  True, False,  True,
       False,  True, False,  True, False,  True, False,  True, False,
        True, False,  True, False,  True, False,  True, False,  True,
       False,  True, False,  True, False,  True, False,  True, False,
        True, False,  True, False,  True, False,  True, False,  True,
       False,  True, False,  True, False,  True, False,  True, False,
        True, False,  True, False,  True, False,  True, False,  True,
       False,  True, False,  True, False,  True, False,  True, False,
        True, False,

In [114]:
df[ df.index % 2== 0 ]

PassengerId  Survived  Pclass  \
0              1         0       3   
2              3         1       3   
4              5         0       3   
6              7         0       1   
8              9         1       3   
..           ...       ...     ...   
882          883         0       3   
884          885         0       3   
886          887         0       2   
888          889         0       3   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
4                             Allen, Mr. William Henry    male  35.0      0   
6                              McCarthy, Mr. Timothy J    male  54.0      0   
8    Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
..                                                 ...     ...   ...    ...   
882                       Dahlberg, Miss. Gerda Ulrika  female  22.0      0   
884                             Sutehall, Mr. Henry Jr    male  25.0      0   
886                              Montvila, Rev. Juozas    male  27.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  new_col  new_col1  
0        0         A/5 21171   7.2500   NaN        S        0         4  
2        0  STON/O2. 3101282   7.9250   NaN        S        0         6  
4        0            373450   8.0500   NaN        S        0         8  
6        0             17463  51.8625   E46        S        0         8  
8        2            347742  11.1333   NaN        S        0        12  
..     ...               ...      ...   ...      ...      ...       ...  
882      0              7552  10.5167   NaN        S        0       886  
884      0   SOTON/OQ 392076   7.0500   NaN        S        0       888  
886      0            211536  13.0000   NaN        S        0       889  
888      2        W./C. 6607  23.4500   NaN        S        0       892  
890      0            370376   7.7500   NaN        Q        0       894  

[446 rows x 14 columns]

In [118]:
df[ (df.index % 2== 0) & (df.index < 100) ]

PassengerId  Survived  Pclass  \
0             1         0       3   
2             3         1       3   
4             5         0       3   
6             7         0       1   
8             9         1       3   
10           11         1       3   
12           13         0       3   
14           15         0       3   
16           17         0       3   
18           19         0       3   
20           21         0       2   
22           23         1       3   
24           25         0       3   
26           27         0       3   
28           29         1       3   
30           31         0       1   
32           33         1       3   
34           35         0       1   
36           37         1       3   
38           39         0       3   
40           41         0       3   
42           43         0       3   
44           45         1       3   
46           47         0       3   
48           49         0       3   
50           51         0       3   
52           53         1       1   
54           55         0       1   
56           57         1       2   
58           59         1       2   
60           61         0       3   
62           63         0       1   
64           65         0       1   
66           67         1       2   
68           69         1       3   
70           71         0       2   
72           73         0       2   
74           75         1       3   
76           77         0       3   
78           79         1       2   
80           81         0       3   
82           83         1       3   
84           85         1       2   
86           87         0       3   
88           89         1       1   
90           91         0       3   
92           93         0       1   
94           95         0       3   
96           97         0       1   
98           99         1       2   

                                                 Name     Sex    Age  SibSp  \
0                             Braund, Mr. Owen Harris    male  22.00      1   
2                              Heikkinen, Miss. Laina  female  26.00      0   
4                            Allen, Mr. William Henry    male  35.00      0   
6                             McCarthy, Mr. Timothy J    male  54.00      0   
8   Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.00      0   
10                    Sandstrom, Miss. Marguerite Rut  female   4.00      1   
12                     Saundercock, Mr. William Henry    male  20.00      0   
14               Vestrom, Miss. Hulda Amanda Adolfina  female  14.00      0   
16                               Rice, Master. Eugene    male   2.00      4   
18  Vander Planke, Mrs. Julius (Emelia Maria Vande...  female  31.00      1   
20                               Fynney, Mr. Joseph J    male  35.00      0   
22                        McGowan, Miss. Anna "Annie"  female  15.00      0   
24                      Palsson, Miss. Torborg Danira  female   8.00      3   
26                            Emir, Mr. Farred Chehab    male    NaN      0   
28                      O'Dwyer, Miss. Ellen "Nellie"  female    NaN      0   
30                           Uruchurtu, Don. Manuel E    male  40.00      0   
32                           Glynn, Miss. Mary Agatha  female    NaN      0   
34                            Meyer, Mr. Edgar Joseph    male  28.00      1   
36                                   Mamee, Mr. Hanna    male    NaN      0   
38                 Vander Planke, Miss. Augusta Maria  female  18.00      2   
40     Ahlin, Mrs. Johan (Johanna Persdotter Larsson)  female  40.00      1   
42                                Kraeff, Mr. Theodor    male    NaN      0   
44                      Devaney, Miss. Margaret Delia  female  19.00      0   
46                                  Lennon, Mr. Denis    male    NaN      1   
48                                Samaan, Mr. Youssef    male    NaN      2   
50                         Panula, Master. Juha Niilo    male 

Easier Solution:

In [116]:
df[0:100:2]

PassengerId  Survived  Pclass  \
0             1         0       3   
2             3         1       3   
4             5         0       3   
6             7         0       1   
8             9         1       3   
10           11         1       3   
12           13         0       3   
14           15         0       3   
16           17         0       3   
18           19         0       3   
20           21         0       2   
22           23         1       3   
24           25         0       3   
26           27         0       3   
28           29         1       3   
30           31         0       1   
32           33         1       3   
34           35         0       1   
36           37         1       3   
38           39         0       3   
40           41         0       3   
42           43         0       3   
44           45         1       3   
46           47         0       3   
48           49         0       3   
50           51         0       3   
52           53         1       1   
54           55         0       1   
56           57         1       2   
58           59         1       2   
60           61         0       3   
62           63         0       1   
64           65         0       1   
66           67         1       2   
68           69         1       3   
70           71         0       2   
72           73         0       2   
74           75         1       3   
76           77         0       3   
78           79         1       2   
80           81         0       3   
82           83         1       3   
84           85         1       2   
86           87         0       3   
88           89         1       1   
90           91         0       3   
92           93         0       1   
94           95         0       3   
96           97         0       1   
98           99         1       2   

                                                 Name     Sex    Age  SibSp  \
0                             Braund, Mr. Owen Harris    male  22.00      1   
2                              Heikkinen, Miss. Laina  female  26.00      0   
4                            Allen, Mr. William Henry    male  35.00      0   
6                             McCarthy, Mr. Timothy J    male  54.00      0   
8   Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.00      0   
10                    Sandstrom, Miss. Marguerite Rut  female   4.00      1   
12                     Saundercock, Mr. William Henry    male  20.00      0   
14               Vestrom, Miss. Hulda Amanda Adolfina  female  14.00      0   
16                               Rice, Master. Eugene    male   2.00      4   
18  Vander Planke, Mrs. Julius (Emelia Maria Vande...  female  31.00      1   
20                               Fynney, Mr. Joseph J    male  35.00      0   
22                        McGowan, Miss. Anna "Annie"  female  15.00      0   
24                      Palsson, Miss. Torborg Danira  female   8.00      3   
26                            Emir, Mr. Farred Chehab    male    NaN      0   
28                      O'Dwyer, Miss. Ellen "Nellie"  female    NaN      0   
30                           Uruchurtu, Don. Manuel E    male  40.00      0   
32                           Glynn, Miss. Mary Agatha  female    NaN      0   
34                            Meyer, Mr. Edgar Joseph    male  28.00      1   
36                                   Mamee, Mr. Hanna    male    NaN      0   
38                 Vander Planke, Miss. Augusta Maria  female  18.00      2   
40     Ahlin, Mrs. Johan (Johanna Persdotter Larsson)  female  40.00      1   
42                                Kraeff, Mr. Theodor    male    NaN      0   
44                      Devaney, Miss. Margaret Delia  female  19.00      0   
46                                  Lennon, Mr. Denis    male    NaN      1   
48                                Samaan, Mr. Youssef    male    NaN      2   
50                         Panula, Master. Juha Niilo    male 

In [119]:
# Give only first two rows

df[0:2]

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   

   Parch     Ticket     Fare Cabin Embarked  new_col  new_col1  
0      0  A/5 21171   7.2500   NaN        S        0         4  
1      0   PC 17599  71.2833   C85        C        0         3

Now, we can also use iloc

In [122]:
df.iloc[0:2]   # iloc excludes the  last element of the range

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   

   Parch     Ticket     Fare Cabin Embarked  new_col  new_col1  
0      0  A/5 21171   7.2500   NaN        S        0         4  
1      0   PC 17599  71.2833   C85        C        0         3

In [123]:
df.loc[0:2]  # loc includes the  last element of the range

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  new_col  new_col1  
0      0         A/5 21171   7.2500   NaN        S        0         4  
1      0          PC 17599  71.2833   C85        C        0         3  
2      0  STON/O2. 3101282   7.9250   NaN        S        0         6

* The loc() function is label based data selecting method which means that we have to pass the name of the row or column which we want to select. This method includes the last element of the range passed in it, unlike iloc(). loc() can accept the boolean data unlike iloc().

* The iloc() function is an indexed-based selecting method which means that we have to pass an integer index in the method to select a specific row/column. This method does not include the last element of the range passed in it unlike loc(). iloc() does not accept the boolean data unlike loc().

In [128]:
# Select first 2 rows and first 3 columns

df.iloc[0:2, ['PassengerId'	,'Survived'	,'Pclass']]

IndexError: ignored

In [129]:
df.iloc[0:2, 0:3]

PassengerId  Survived  Pclass
0            1         0       3
1            2         1       1

In [130]:
# Select first 2 rows and first 3 columns

df.loc[0:2, ['PassengerId'	,'Survived'	,'Pclass'] ]

PassengerId  Survived  Pclass
0            1         0       3
1            2         1       1
2            3         1       3

iloc only works with inbuilt indexes whereas loc works with inbuilt indexes for rows and named indexes for columns.

inbuilt indexes -> 0,1,2,3

default/ named indexes -> PassengerId, Pclass, Sex

In [133]:
df.loc[0:2, 0:3]

TypeError: ignored

For loc, slicing can be done for rows and not for columns. This is because for rows, inbuilt index is taken always as there is no named index unlike that of column where named index is present.